city_metrix/layers/building_classifier.py

```diff
@@ -0,0 +1,181 @@
+            # Add other mappings as needed
+        }
+        reclassified_esa = xr.apply_ufunc(
+            np.vectorize(lambda x: reclass_map.get(x, x)),
+            esa_world_cover,
+            vectorize=True
+        )
+
+        reclassified_esa = reclassified_esa.rio.write_crs(esa_world_cover.rio.crs, inplace=True)
+
+        esa_1m = reclassified_esa.rio.reproject(
+            dst_crs=crs,
+            resolution=1,
+            resampling=Resampling.nearest
+        )
+
+        # Read ULU land cover, filter to city, select lulc band
+        ulu_lulc = UrbanLandUse(band='lulc').get_data(bbox)
+        ulu_roads = UrbanLandUse(band='road').get_data(bbox)
+        # Create road mask of 50
+        # Typical threshold for creating road mask 
+        road_mask = ulu_roads >= 50
+        ulu_lulc = ulu_lulc.where(~road_mask, 6)
+        # 1-Non-residential: 0 (open space), 1 (non-res)
+        # 2-Residential: 2 (Atomistic), 3 (Informal), 4 (Formal), 5 (Housing project)
+        # 3-Roads: 6 (Roads)
+        mapping = {0: 1, 1: 1, 2: 2, 3: 2, 4: 2, 5: 2, 6: 3}
+        for from_val, to_val in mapping.items():
+            ulu_lulc = ulu_lulc.where(ulu_lulc != from_val, to_val)
+        
+        # 1-Non-residential as default
+        ulu_lulc_1m = ulu_lulc.rio.reproject(
+            dst_crs=crs,
+            shape=esa_1m.shape,
+            resampling=Resampling.nearest,
+            nodata=1
+        )
+
+        # ANBH is the average height of the built surfaces, USE THIS
+        # AGBH is the amount of built cubic meters per surface unit in the cell
+        # https://ghsl.jrc.ec.europa.eu/ghs_buH2023.php
+        anbh = (ee.ImageCollection("projects/wri-datalab/GHSL/GHS-BUILT-H-ANBH_R2023A")
+                .filterBounds(ee.Geometry.BBox(*bbox))
+                .select('b1')
+                .mosaic()
+                )
+        ds = xr.open_dataset(
+            ee.ImageCollection(anbh),
+            engine='ee',
+            scale=100,
+            crs=crs,
+            geometry=ee.Geometry.Rectangle(*bbox)
+        )
+        anbh_data = ds.b1.compute()
+        # get in rioxarray format
+        anbh_data = anbh_data.squeeze("time").transpose("Y", "X").rename({'X': 'x', 'Y': 'y'})
+        
+        anbh_1m = anbh_data.rio.reproject(
+            dst_crs=crs,
+            shape=esa_1m.shape,
+            resampling=Resampling.nearest,
+            nodata=0
+        )
+
+        return buildings_sample, esa_1m, ulu_lulc_1m, anbh_1m, crs
+    
+
+    # Extract values to buildings as coverage fractions
+    # Extract average of pixel values to buildings
+    # Reproject to local state plane and calculate area
+    def calc_majority_ULU_mean_ANBH_area(self, row, building_sample_1m, id_col, ulu_lulc_1m, anbh_1m):
+        mask = building_sample_1m == row[id_col]
+        masked_ulu = ulu_lulc_1m.values[mask]
+        
+        # Extract values to buildings as coverage fractions
```

Comment:
what are the coverage fractions? the coverage of each pixel by a building? I'm a little confused on how the code below is accomplishing this. It might be better to use less complex indexing, it makes the code hard to follow.